subiquity/server/ubuntu_drivers.py

```diff
@@ -168,6 +173,145 @@ async def list_oem(self, root_dir: str, context) -> List[str]:
         return self._oem_metapackages_from_output(result.stdout.decode("utf-8"))
 
 
+class UbuntuDriversFakePCIDevicesInterface(UbuntuDriversInterface):
+    """An implementation of ubuntu-drivers that wraps the calls with
+    the umockdev wrapper script.
+
+
+    Requires an online install to download umockdev packages or
+    a modified ISOs with the packages added to the pool.
+    """
+
+    def __init__(self, app, gpgpu: bool) -> None:
+        super().__init__(app, gpgpu)
+
+        # PCI devices can be passed on the kernel command line as a comma
+        # separated list:
+        #  subiquity-fake-pci-devices=pci:v00001234d00001234,pci:v00005678d00005678
+        pcis = app.opts.kernel_cmdline.get("subiquity-fake-pci-devices")
+        devs = [self.modalias_to_config(d) for d in pcis.split(",") if d != ""]
+        self.dev_config = {"devices": devs}
+        log.debug(f"writing umockdev_config: {self.dev_config}")
+
+        # write config to live environment
+        self.dev_config_path = "/tmp/umockdev_config.yaml"
+        write_file(self.dev_config_path, yaml.safe_dump(self.dev_config), mode=0o777)
+
+        prefix: list[str] = [
+            "subiquity-umockdev-wrapper",  # vendored in system_scripts
+            "--config",
+            self.dev_config_path,
+            "--",  # Don't let wrapper consume ubuntu-drivers feature flags
+        ]
+
+        self.sys_env = system_scripts_env()
+
+        self.pre_req_cmd = [
+            "apt-get",
+            "install",
+            "-y",
+            "umockdev",
+            "gir1.2-umockdev-1.0",
+        ]
```

Comment:
there seems to be an race condition / apt conflict here, I had to launch 3 times to see this command pass.

Reply:
Thanks for pointing this out. Pushed a new commit to fix this by waiting for the apt lock with `-oDPkg::Lock::Timeout=-1`.